### PR TITLE
Network updates

### PR DESCRIPTION
**Pull request by lenala**

### CODE REVIEW

File: src/ResourceManagement/Network/AzureReachabilityReportImpl.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        ///GENMHASH:AF5B92F7D602ADD6835472AD2633A663:7A31B9AAF5466328BC49ECCDA3102A92
+        public AzureReachabilityReportParameters AzureReachabilityReportParameters()
+        {
+            return parameters;
+        }
+
+        ///GENMHASH:637F0A3522F2C635C23E54FAAD79CBEA:4891FA97A857767ACE1138F0F8DDD55F
+        public override async Task<IAzureReachabilityReport> ExecuteAsync(CancellationToken cancellationToken = new CancellationToken(), bool multiThreaded = true)
```
Comment (hovsepm) [inline src/ResourceManagement/Network/AzureReachabilityReportImpl.cs]: >CancellationToken [](start = 116, length = 17)

should be default(CancellationToken)

File: src/ResourceManagement/Network/AzureReachabilityReportImpl.cs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        {
+            this.inner =
+                await parent.Manager.Inner.NetworkWatchers.GetAzureReachabilityReportAsync(parent.ResourceGroupName,
+                    parent.Name, parameters);
```
Comment (hovsepm) [inline src/ResourceManagement/Network/AzureReachabilityReportImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/ConnectionMonitorImpl.cs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            if (createParameters.Source == null) {
+                createParameters.Source = new ConnectionMonitorSource();
+            }
+            return createParameters.Source;
+        }
+
+        ///GENMHASH:5AD91481A0966B059A478CD4E9DD9466:754810EA5144700BBB078C6E55E8C153
+        protected override async Task<Models.ConnectionMonitorResultInner> GetInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.client.GetAsync(parent.ResourceGroupName, parent.Name, Name);
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectionMonitorImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/ConnectionMonitorImpl.cs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        public bool AutoStart()
+        {
+            return Inner.AutoStart.GetValueOrDefault();
+        }
+
+        ///GENMHASH:0202A00A1DCF248D2647DBDBEF2CA865:228A9A3D68EE8BB5ABA520AE16356B8A
+        public override async Task<Microsoft.Azure.Management.Network.Fluent.IConnectionMonitor> CreateResourceAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            SetInner(await this.client.CreateOrUpdateAsync(parent.ResourceGroupName, parent.Name, this.Name,
+                createParameters));
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectionMonitorImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/ConnectionMonitorImpl.cs [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        public IConnectionMonitorQueryResult Query()
+        {
+            return Extensions.Synchronize(() => QueryAsync());
+        }
+
+        ///GENMHASH:A409C3D3741509AF429A05127B5E4382:58F7FC16B9BBD602D0D4676357994540
+        public  async Task<Microsoft.Azure.Management.Network.Fluent.IConnectionMonitorQueryResult> QueryAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return new ConnectionMonitorQueryResultImpl(await this.client.QueryAsync(parent.ResourceGroupName, parent.Name, Name));
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectionMonitorImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/ConnectionMonitorQueryResultImpl.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        public ConnectionMonitorSourceStatus SourceStatus()
+        {
+            return Inner.SourceStatus;
+        }
+
+        ///GENMHASH:BE1EA957D0BFB4FE19638B8E97D6071D:3AE0E4C8B5F32B836364999BCF8BD994
+        public IReadOnlyList<Models.ConnectionStateSnapshot> States()
+        {
+            //$ return Collections.UnmodifiableList(Inner.States());
+
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectionMonitorQueryResultImpl.cs]: this does not look like implemented code.

File: src/ResourceManagement/Network/ConnectionMonitorsImpl.cs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            Extensions.Synchronize(() => DeleteByIdAsync(id));
+        }
+
+
+        ///GENMHASH:4D33A73A344E127F784620E76B686786:B016E9F1E898A15E4995A241E8FB38A5
+        public override async Task DeleteByIdAsync(string id, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            ResourceId resourceId = ResourceId.FromString(id);
+            await this.innerCollection.DeleteAsync(resourceId.ResourceGroupName, resourceId.Parent.Name, resourceId.Name);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectionMonitorsImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/ConnectionMonitorsImpl.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        {
+            Extensions.Synchronize(() => DeleteByNameAsync(name));
+
+        }
+
+        ///GENMHASH:971272FEE209B8A9A552B92179C1F926:E5B162BD6005B3BA236ADFAE6CA0A4CF
+        public async Task DeleteByNameAsync(string name, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            await this.innerCollection.DeleteAsync(parent.ResourceGroupName, parent.Name, name);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectionMonitorsImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/ConnectionMonitorsImpl.cs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        public IConnectionMonitor GetByName(string name)
+        {
+            return Extensions.Synchronize(() => GetByNameAsync(name));
+        }
+
+        ///GENMHASH:885F10CFCF9E6A9547B0702B4BBD8C9E:2AE56F064529C0748B82B77D0D690DFC
+        public async Task<Microsoft.Azure.Management.Network.Fluent.IConnectionMonitor> GetByNameAsync(string name, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await innerCollection.GetAsync(parent.ResourceGroupName, parent.Name, name);
+            return (WrapModel(inner));
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectionMonitorsImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/ConnectivityCheckImpl.cs [hunk 10]
```diff
@@ -148,5 +148,14 @@ public ConnectivityCheckImpl ToDestinationAddress(string address)
                 .CheckConnectivityAsync(parent.ResourceGroupName, parent.Name, parameters);
             return this;
         }
+
+        ///GENMHASH:BB86B1F07AADE33841AA4762FF3CC20E:B387E3DD8624B2499C0BDB4DDC202E72
+        public ConnectivityCheckImpl WithProtocol(Protocol protocol)
+        {
+            //$ parameters.WithProtocol(protocol);
+            //$ return this;
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ConnectivityCheckImpl.cs]: not fully implemented

File: src/ResourceManagement/Network/DdosProtectionPlanImpl.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        protected override async Task<Models.DdosProtectionPlanInner> GetInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.Inner.DdosProtectionPlans.GetAsync(this.ResourceGroupName, this.Name, cancellationToken);
+        }
+
+        ///GENMHASH:0202A00A1DCF248D2647DBDBEF2CA865:575791C5B65702BB0207EBB86C505DCF
+        public override async Task<Microsoft.Azure.Management.Network.Fluent.IDdosProtectionPlan> CreateResourceAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            //$ return this.Manager().Inner.DdosProtectionPlans().CreateOrUpdateAsync(resourceGroupName(), name(), Inner)
+            //$ .Map(innerToFluentMap(this));
```
Comment (hovsepm) [inline src/ResourceManagement/Network/DdosProtectionPlanImpl.cs]: not implemented

File: src/ResourceManagement/Network/DdosProtectionPlanImpl.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        public string ResourceGuid()
+        {
+            return Inner.ResourceGuid;
+        }
+
+        ///GENMHASH:3432D5F9AA6D1AC31DF96CB75325C37A:A14ACA04E5E1D457B8AC97821B48DDD6
+        public IReadOnlyList<SubResource> VirtualNetworks()
+        {
+            //$ return Collections.UnmodifiableList(Inner.VirtualNetworks());
+
```
Comment (hovsepm) [inline src/ResourceManagement/Network/DdosProtectionPlanImpl.cs]: not implemented.

File: src/ResourceManagement/Network/ExpressRouteCircuitImpl.cs [hunk 13]
```diff
@@ -147,6 +147,12 @@ public ExpressRouteCircuitImpl WithSku(ExpressRouteCircuitSkuType sku)
             return this;
         }
 
+        ///GENMHASH:0263B1F6C1D2EA8755C6E28644653147:B6AF4318DCA2628C503284303A321B0F
+        protected override async Task<Models.ExpressRouteCircuitInner> ApplyTagsToInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.Inner.ExpressRouteCircuits.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ExpressRouteCircuitImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/GroupableParentResourceWithTags.cs [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+//        {
+//            base.WithTag(key, value);
+//            return this as IAppliableWithTags<IFluentResourceT>;
+//        }
+//
+//        public IAppliableWithTags<IFluentResourceT> WithTags(IDictionary<string, string> tags)
+//        {
+//            base.WithTags(tags);
+//            return this as IAppliableWithTags<IFluentResourceT>;
+//        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/GroupableParentResourceWithTags.cs]: uncomment or delete comments.

File: src/ResourceManagement/Network/LocalNetworkGatewayImpl.cs [hunk 15]
```diff
@@ -41,6 +45,27 @@ public ISet<string> AddressSpaces()
...
+        public ILocalNetworkGateway ApplyTags()
+        {
+            return Extensions.Synchronize(() => ApplyTagsAsync());
+        }
+
+        ///GENMHASH:6B8BA63027964E06F44A927837B450A0:73A3E3DE52EE39671B79DFFC66BC8B20
+        public async Task<Microsoft.Azure.Management.Network.Fluent.ILocalNetworkGateway> ApplyTagsAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await Manager.Inner.LocalNetworkGateways.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
+            SetInner(inner);
```
Comment (hovsepm) [inline src/ResourceManagement/Network/LocalNetworkGatewayImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/NetworkImpl.cs [hunk 16]
```diff
@@ -201,6 +204,18 @@ internal SubnetImpl UpdateSubnet(string name)
         ///GENMHASH:359B78C1848B4A526D723F29D8C8C558:A394B5B2A4C946983BF7F8DE2DAA697E
         protected async override Task<VirtualNetworkInner> CreateInnerAsync(CancellationToken cancellationToken)
         {
+        //$ if (ddosProtectionPlanCreatable != null && this.TaskResult(ddosProtectionPlanCreatable.Key()) != null) {
+                    //$ DdosProtectionPlan ddosProtectionPlan = this.<DdosProtectionPlan>taskResult(ddosProtectionPlanCreatable.Key());
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkImpl.cs]: not implemented.

File: src/ResourceManagement/Network/NetworkSecurityGroupImpl.cs [hunk 17]
```diff
@@ -108,6 +108,11 @@ internal NetworkSecurityRuleImpl UpdateRule(string name)
             return (NetworkSecurityRuleImpl)rule;
         }
 
+        ///GENMHASH:0263B1F6C1D2EA8755C6E28644653147:00DFE9718DCBB33CB69317A9896AE9D7
+        protected override async Task<Models.NetworkSecurityGroupInner> ApplyTagsToInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.Inner.NetworkSecurityGroups.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkSecurityGroupImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/NetworkSecurityRuleImpl.cs [hunk 18]
```diff
@@ -304,5 +324,16 @@ internal NetworkSecurityGroupImpl Attach()
             return Parent.WithRule(this);
         }
         #endregion
+
+        ///GENMHASH:A6BD27BDA73F3A8D110E59E9BC1BF594:B3E2FF55E7190758DABFDE8C609EEECA
+        public NetworkSecurityRuleImpl WithSourceApplicationSecurityGroup(string id)
+        {
+
+            //$ sourceAsgs.Put(id, new ApplicationSecurityGroupInner().WithId(id));
+            Inner.SourceAddressPrefix = null;
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkSecurityRuleImpl.cs]: not implemented?

File: src/ResourceManagement/Network/NetworkSecurityRuleImpl.cs [hunk 19]
```diff
@@ -285,7 +296,16 @@ private NetworkSecurityRuleImpl WithDirection(SecurityRuleDirection direction)
             Inner.Direction = direction;
             return this;
         }
+        ///GENMHASH:2B7FF25B14EF62E55CC8805B824E0366:11B5B307793C3F9FF1E037CD73A12C40
+        public NetworkSecurityRuleImpl WithDestinationApplicationSecurityGroup(string id)
+        {
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkSecurityRuleImpl.cs]: not implemented.

File: src/ResourceManagement/Network/NetworkWatcherImpl.cs [hunk 20]
```diff
@@ -133,5 +158,26 @@ public TopologyImpl Topology()
...
+        public INetworkWatcher ApplyTags()
+        {
+            return Extensions.Synchronize(() => ApplyTagsAsync());
+        }
+
+        ///GENMHASH:6B8BA63027964E06F44A927837B450A0:73A3E3DE52EE39671B79DFFC66BC8B20
+        public async Task<Microsoft.Azure.Management.Network.Fluent.INetworkWatcher> ApplyTagsAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await Manager.Inner.NetworkWatchers.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
+            SetInner(inner);
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkWatcherImpl.cs]: missing cancellationToken parameter.

File: src/ResourceManagement/Network/NetworkWatcherImpl.cs [hunk 21]
```diff
@@ -48,6 +50,21 @@ public NextHopImpl NextHop()
             return new NextHopImpl(this);
         }
 
+        ///GENMHASH:E57B0EEDF58A342B01A5F4CF76030BAC:4CDB58B16C3B63EB3194D9C1D695E7C7
+        public AvailableProvidersImpl AvailableProviders()
+        {
+            //$ return new AvailableProvidersImpl(this);
+
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkWatcherImpl.cs]: not implemented

File: src/ResourceManagement/Network/NetworkWatcherImpl.cs [hunk 22]
```diff
@@ -48,6 +50,21 @@ public NextHopImpl NextHop()
...
+            return null;
+        }
+
+        ///GENMHASH:1470B429A1D139BF4EBFE36C097746F5:D5E9DFF45F32337A2A4B1D2D29D7F031
+        public AzureReachabilityReportImpl AzureReachabilityReport()
+        {
+            //$ return new AzureReachabilityReportImpl(this);
+
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkWatcherImpl.cs]: not implemented

File: src/ResourceManagement/Network/NetworkWatcherImpl.cs [hunk 23]
```diff
@@ -61,6 +78,14 @@ public ConnectivityCheckImpl CheckConnectivity()
             return new ConnectivityCheckImpl(this);
         }
 
+        ///GENMHASH:E3A05E297410D043611B64D370D8BA54:094A9222342E6F539E7A7FB5FD5A5920
+        public ConnectionMonitorsImpl ConnectionMonitors()
+        {
+            //$ return connectionMonitors;
+
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkWatcherImpl.cs]: not implemented

File: src/ResourceManagement/Network/PublicIPAddressImpl.cs [hunk 24]
```diff
@@ -287,5 +288,26 @@ internal INicIPConfiguration GetAssignedNetworkInterfaceIPConfiguration()
...
+        ///GENMHASH:BCABB5578B0BD7DC8F8C22F4769FD3DE:02984D22C1D2E484D62F2595E7B0E86C
+        public IPublicIPAddress ApplyTags()
+        {
+            return Extensions.Synchronize(() => ApplyTagsAsync());
+        }
+
+        ///GENMHASH:6B8BA63027964E06F44A927837B450A0:73A3E3DE52EE39671B79DFFC66BC8B20
+        public async Task<Microsoft.Azure.Management.Network.Fluent.IPublicIPAddress> ApplyTagsAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await Manager.Inner.PublicIPAddresses.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
```
Comment (hovsepm) [inline src/ResourceManagement/Network/PublicIPAddressImpl.cs]: missing canellationToken parameter

File: src/ResourceManagement/Network/Domain/InterfaceImpl/VirtualNetworkGatewayConnectionImpl.cs [hunk 25]
```diff
@@ -332,5 +332,21 @@ VirtualNetworkGatewayConnection.Definition.IWithSharedKey VirtualNetworkGatewayC
         {
             return this.WithSecondVirtualNetworkGateway(virtualNetworkGateway2) as VirtualNetworkGatewayConnection.Definition.IWithSharedKey;
         }
+
+        public IAppliableWithTags<IVirtualNetworkGatewayConnection> WithoutTag(string key)
+        {
+            return this.WithoutTag(key) as IAppliableWithTags<IVirtualNetworkGatewayConnection>;
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/Domain/InterfaceImpl/VirtualNetworkGatewayConnectionImpl.cs]: remove all `as` casts. They hide real errors.

File: src/ResourceManagement/Network/Domain/InterfaceImpl/PublicIpAddressImpl.cs [hunk 26]
```diff
@@ -306,5 +312,22 @@ PublicIPAddress.Definition.IWithCreate PublicIPAddress.Definition.IWithSku.WithS
         {
             return this.WithSku(skuType) as PublicIPAddress.Definition.IWithCreate;
         }
+
+
+        public IAppliableWithTags<IPublicIPAddress> WithoutTag(string key)
+        {
+            return this.WithoutTag(key) as IAppliableWithTags<IPublicIPAddress>;
```
Comment (hovsepm) [inline src/ResourceManagement/Network/Domain/InterfaceImpl/PublicIpAddressImpl.cs]: >as IAppliableWithTags<IPublicIPAddress> [](start = 40, length = 39)

remove `as` cast in this file.

File: src/ResourceManagement/Network/Domain/InterfaceImpl/NetworkWatcherImpl.cs [hunk 27]
```diff
@@ -39,16 +43,20 @@ NextHop.Definition.IWithTargetResource Microsoft.Azure.Management.Network.Fluent
         /// </summary>
         /// <param name="vmId">ID of the target VM.</param>
         /// <return>The configured and effective security group rules on the specified VM.</return>
-        async Task<Microsoft.Azure.Management.Network.Fluent.ISecurityGroupView> Microsoft.Azure.Management.Network.Fluent.INetworkWatcher.GetSecurityGroupViewAsync(string vmId, CancellationToken cancellationToken)
+        async Task<Microsoft.Azure.Management.Network.Fluent.ISecurityGroupView>
+            Microsoft.Azure.Management.Network.Fluent.INetworkWatcher.GetSecurityGroupViewAsync(string vmId,
+                CancellationToken cancellationToken)
         {
-            return await this.GetSecurityGroupViewAsync(vmId, cancellationToken) as Microsoft.Azure.Management.Network.Fluent.ISecurityGroupView;
+            return await this.GetSecurityGroupViewAsync(vmId, cancellationToken) as
```
Comment (hovsepm) [inline src/ResourceManagement/Network/Domain/InterfaceImpl/NetworkWatcherImpl.cs]: >as [](start = 81, length = 2)

remove `as` cast in this file.

File: src/ResourceManagement/Network/VirtualNetworkGatewayImpl.cs [hunk 28]
```diff
@@ -492,5 +493,10 @@ private VirtualNetworkGatewayIPConfigurationImpl DefineIPConfiguration(string na
             };
             return new VirtualNetworkGatewayIPConfigurationImpl(inner, this);
         }
+
+        public IUpdateWithTags<IVirtualNetworkGateway> UpdateTags()
+        {
+            throw new System.NotImplementedException();
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/VirtualNetworkGatewayImpl.cs]: not implemented? should this be an overriden method then?

File: src/ResourceManagement/Network/VirtualNetworkGatewayConnectionImpl.cs [hunk 29]
```diff
@@ -33,9 +33,25 @@ internal partial class VirtualNetworkGatewayConnectionImpl :
...
+        ///GENMHASH:BCABB5578B0BD7DC8F8C22F4769FD3DE:02984D22C1D2E484D62F2595E7B0E86C
+        public IVirtualNetworkGatewayConnection ApplyTags()
+        {
+            return Extensions.Synchronize(() => ApplyTagsAsync());
+        }
+
+        ///GENMHASH:6B8BA63027964E06F44A927837B450A0:BA32A105329DC11B026CC971C7538262
+        public async Task<Microsoft.Azure.Management.Network.Fluent.IVirtualNetworkGatewayConnection> ApplyTagsAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            await Manager.Inner.VirtualNetworkGatewayConnections.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
```
Comment (hovsepm) [inline src/ResourceManagement/Network/VirtualNetworkGatewayConnectionImpl.cs]: missing cancellationToken

File: src/ResourceManagement/Network/VirtualNetworkGatewayConnectionImpl.cs [hunk 30]
```diff
@@ -274,16 +296,45 @@ public IReadOnlyCollection<TunnelConnectionHealth> TunnelConnectionStatus()
...
+//        {
+//            base.WithTags(tags);
+//            return this;
+//        }
+//
+//        public VirtualNetworkGatewayConnectionImpl WithTag(string key, string value)
+//        {
+//            base.WithTag(key, value);
+//            return this;
+//        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/VirtualNetworkGatewayConnectionImpl.cs]: remove commented code.

File: src/ResourceManagement/Network/VirtualNetworkGatewayConnectionImpl.cs [hunk 31]
```diff
@@ -274,16 +296,45 @@ public IReadOnlyCollection<TunnelConnectionHealth> TunnelConnectionStatus()
...
+        }
+
+        IUpdate IUpdateWithTags<IUpdate>.WithTag(string key, string value)
+        {
+            throw new System.NotImplementedException();
+        }
+
+        IUpdate IUpdateWithTags<IUpdate>.WithoutTag(string key)
+        {
+            throw new System.NotImplementedException();
```
Comment (hovsepm) [inline src/ResourceManagement/Network/VirtualNetworkGatewayConnectionImpl.cs]: >NotImplementedException [](start = 29, length = 23)

If this is by design then change all not implemented exceptions to not supported exception.

File: src/ResourceManagement/Network/RouteTableImpl.cs [hunk 32]
```diff
@@ -132,6 +135,11 @@ internal RouteTableImpl WithRoute(RouteImpl route)
             return this;
         }
 
+        ///GENMHASH:0263B1F6C1D2EA8755C6E28644653147:9D293E1AA04D2C3CD8EDA9D8EB8FEDDD
+        protected override async Task<Models.RouteTableInner> ApplyTagsToInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.Inner.RouteTables.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/RouteTableImpl.cs]: cancellationToken

File: src/ResourceManagement/Network/RouteTableImpl.cs [hunk 33]
```diff
@@ -165,5 +173,7 @@ public IUpdate WithoutRoute(string name)
             routes.Remove(name);
             return this;
         }
+
+        public bool IsBgpRoutePropagationDisabled { get; }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/RouteTableImpl.cs]: >IsBgpRoutePropagationDisabled [](start = 20, length = 29)

this does not seems to be implemented correctly. who sets the value?

File: src/ResourceManagement/Network/RouteFiltersImpl.cs [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+        ///GENMHASH:8ACFB0E23F5F24AD384313679B65F404:AD7C28D26EC1F237B93E54AD31899691
+        public RouteFilterImpl Define(string name)
+        {
+            return WrapModel(name);
+        }
+
+        protected override async Task<RouteFilterInner> GetInnerByGroupAsync(string groupName, string name, CancellationToken cancellationToken)
+        {
+            return await Inner.GetAsync(groupName, name);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/RouteFiltersImpl.cs]: cancellationToken

File: src/ResourceManagement/Network/RouteFilterImpl.cs [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+        protected override async Task<Models.RouteFilterInner> CreateInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.Inner.RouteFilters.CreateOrUpdateAsync(ResourceGroupName, Name, Inner, cancellationToken);
+        }
+
+        ///GENMHASH:5AD91481A0966B059A478CD4E9DD9466:AAC56B9AF93DB5A1FE5E71C6B5316254
+        protected override async Task<Models.RouteFilterInner> GetInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.Inner.RouteFilters.GetAsync(this.ResourceGroupName, this.Name);
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/RouteFilterImpl.cs]: cancellationToken

File: src/ResourceManagement/Network/NetworkImpl.cs [hunk 36]
```diff
@@ -201,7 +205,13 @@ internal SubnetImpl UpdateSubnet(string name)
...
         {
-            return await Manager.Inner.VirtualNetworks.CreateOrUpdateAsync(ResourceGroupName, Name, Inner, cancellationToken);
+            if (ddosProtectionPlanCreatable != null)
+            {
+                var ddosProtectionPlan = this.CreatedResource(ddosProtectionPlanCreatable.Key);
+                WithExistingDdosProtectionPlan(ddosProtectionPlan.Id);   
+            }
+            ddosProtectionPlanCreatable = null;
+            return await this.Manager.Inner.VirtualNetworks.CreateOrUpdateAsync(this.ResourceGroupName, this.Name, this.Inner);
         }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkImpl.cs]: cancellationToken

Comment (hovsepm) [inline src/ResourceManagement/Network/NetworkImpl.cs]: cancellationToken

File: src/ResourceManagement/Network/GroupableParentResourceWithTags.cs [hunk 37]
```diff
@@ -0,0 +1,84 @@
...
+
+    {
+        protected GroupableParentResourceWithTags(string name, InnerResourceT innerObject, ManagerT manager) : base(
+            name, innerObject, manager)
+        {
+        }
+
+        public IUpdateWithTags<IFluentResourceT> UpdateTags()
+        {
+            return this as IUpdateWithTags<IFluentResourceT>;
```
Comment (hovsepm) [inline src/ResourceManagement/Network/GroupableParentResourceWithTags.cs]: >as IUpdateWithTags<IFluentResourceT>; [](start = 24, length = 37)

change `as` cast to explicit cast

File: src/ResourceManagement/Network/GroupableParentResourceWithTags.cs [hunk 38]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        protected abstract Task<InnerResourceT> ApplyTagsToInnerAsync(CancellationToken cancellationToken = default(CancellationToken));
+
+        public async Task<IFluentResourceT> ApplyTagsAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var inner = await ApplyTagsToInnerAsync(cancellationToken);
+            SetInner(inner);
+            return this as FluentResourceT;
+        }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/GroupableParentResourceWithTags.cs]: make explicit cast instead

File: src/ResourceManagement/Network/AvailableProvidersImpl.cs [hunk 39]
```diff
@@ -0,0 +1,133 @@
...
+        ///GENMHASH:FD5D5A8D6904B467321E345BE1FA424E:8AB87020DE6C711CD971F3D80C33DD83
+        public INetworkWatcher Parent()
+        {
+            return parent;
+        }
+
+        ///GENMHASH:02D4291678B9DD815B24067695A568FE:CBB7D90388387B638B8F355DFCA9DB53
+        public IReadOnlyDictionary<string,Models.AvailableProvidersListCountry> ProvidersByCountry()
+        {
+            return this.providersByCountry as IReadOnlyDictionary<string, AvailableProvidersListCountry>;
```
Comment (hovsepm) [inline src/ResourceManagement/Network/AvailableProvidersImpl.cs]: >as IReadOnlyDictionary< [](start = 43, length = 23)

change `as` cast

File: src/ResourceManagement/Network/ApplicationGatewayImpl.cs [hunk 40]
```diff
@@ -375,7 +375,11 @@ public IUpdate WithoutPublicIPAddress()
             return WithoutPublicFrontend();
         }
 
-
+        ///GENMHASH:0263B1F6C1D2EA8755C6E28644653147:C2A348673547AB17A3E43B0CDED4C1ED
+        protected override async Task<Models.ApplicationGatewayInner> ApplyTagsToInnerAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return await this.Manager.Inner.ApplicationGateways.UpdateTagsAsync(ResourceGroupName, Name, Inner.Tags);
```
Comment (hovsepm) [inline src/ResourceManagement/Network/ApplicationGatewayImpl.cs]: >Inner [](start = 105, length = 5)

cancellationToken

File: src/ResourceManagement/Network/RouteTableImpl.cs [hunk 41]
```diff
@@ -158,12 +166,27 @@ public RouteImpl UpdateRoute(string name)
...
+        }
 
         ///GENMHASH:7E8C2CF692FADDB1CA4FDD208175D2BD:DA4ECB14ABD20F728F32A3B6D3951691
         public IUpdate WithoutRoute(string name)
         {
             routes.Remove(name);
             return this;
         }
+
+        public bool IsBgpRoutePropagationDisabled { get; }
```
Comment (hovsepm) [inline src/ResourceManagement/Network/RouteTableImpl.cs]: >IsBgpRoutePropagationDisabled [](start = 20, length = 29)

This is still not implemented.

Comment (lenala) [reply src/ResourceManagement/Network/RouteTableImpl.cs]: Done.

**Pull request merged**